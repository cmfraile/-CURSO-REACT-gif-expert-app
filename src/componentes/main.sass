.fondodefault
    background-size: cover
    background-repeat: no-repeat
    background-position: center
.gifgrid
    display: flex
    justify-content: center
    align-items: center
    flex-direction: row
    flex-wrap: wrap
    min-width: 100%
    margin: 50px 0px
    h3
        min-width: 100%
    .ggimg
        position: relative
        min-width: 200px
        max-width: 200px
        min-height: 200px
        max-height: 200px
        margin: 15px
        border-radius: 20px
        .titulo
            position: absolute
            bottom: -15px
            right: -15px
            padding: 3px
            background-color: rgba(255,255,255,1)
            border: 2px solid black
            border-radius: 10px
