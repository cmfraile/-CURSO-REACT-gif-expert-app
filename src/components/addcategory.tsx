import { useState } from "react";
import propTypes from 'prop-types';

//export const AddCategory = ({sC}:{sC?:React.Dispatch<SetStateAction<string[]|[]>>}) => {
interface ACface {onNewCategory:any,cHK:string[]|[]};
export const AddCategory = ({onNewCategory,cHK}:ACface) => {

    const [inputValue,setInputValue] = useState<string>('')

    return(
        <form   aria-label='form'
                onSubmit={(e) => {
            //console.log('Hola mundo desde el submit');
            e.preventDefault();
            
            //VALIDADORES:
            if(inputValue.trim().length <= 2){return}
            if(cHK.map(x => x.toUpperCase()).includes(inputValue.trim().toUpperCase())){return};
            
            //sC((cat) => [inputValue,...cat]);
            onNewCategory(inputValue.trim())
            setInputValue('');
        }}>
            <input  type="text"
                    value={inputValue}
                    aria-label='inputcat'
                    placeholder="buscar gifs"
                    onChange={(e) => {setInputValue(e.target.value)}}
            />
        </form>
    )
}

AddCategory.propTypes = {
    onNewCategory:propTypes.func.isRequired
}
