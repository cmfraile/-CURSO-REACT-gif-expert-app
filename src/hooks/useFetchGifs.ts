import { useEffect, useState } from "react";
import { getGifs } from '../helpers/getgifs';

export const useFetchGifs = (categoria:string) => {

    const [images,setImages] = useState<{id:string,title:string,url:string}[]|[]>([]);
    const [isLoading,setIsLoading] = useState(true);

    useEffect(() => {
        getGifs(categoria).then(setImages);
        setIsLoading(false);
    },[]);
    return{images,isLoading}

}