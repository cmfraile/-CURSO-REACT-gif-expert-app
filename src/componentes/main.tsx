import { useState } from 'react';
import './main.sass';

export const main = () => {

    const [ categorias , setCategorias ] = useState();

    return(
        <>
            {/*Titulo*/}
            <h1>Hola mundo</h1>

            {/*Input*/}
            <ol>
                {['lorem','ipsum','dolor'].map(valor => {
                    return <li key={valor}>{valor}</li>
                })}
            </ol>

            {/*Lista de gifs*/}
                {/*Gif cards*/}
        </>
    )
}