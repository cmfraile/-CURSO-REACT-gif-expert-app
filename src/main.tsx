import React from 'react'
import ReactDOM from 'react-dom/client'
import './main.sass'

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
    <h1>GifExpertApp</h1>
  </React.StrictMode>
)
