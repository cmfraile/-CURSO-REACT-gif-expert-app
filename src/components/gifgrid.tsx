import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

export const GifGrid = ({categoria}:{categoria:string}) => {

    const { images, isLoading } = useFetchGifs(categoria);

    return(

        <div className="gifgrid">
            <h3 aria-label='title' >{categoria}</h3>
            <h2 style={{display:(isLoading) ? '' : 'none'}}>Cargando...</h2>
            {images.map(({id,title,url}) => (
            <div
                aria-label='imgcard'    
                key={id}
                className="ggimg fondodefault"
                style={{backgroundImage:`url(${url})`}}
            >
            <div className="titulo" style={{display:(title) ? '' : 'none' }} >{title}</div>
            </div>
            ))}
        </div>
    )
}

GifGrid.propTypes = { categoria : PropTypes.string.isRequired };