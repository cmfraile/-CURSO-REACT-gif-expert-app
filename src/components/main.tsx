import './main.sass';
import { useState } from 'react';
import { v4 } from 'uuid';
import { AddCategory } from './addcategory';
import { GifGrid } from './gifgrid';

export const main = () => {

    const [ categoriasHook , setCategorias ] = useState<string[]|[]>([]);

    return(
        <>
            <h1>Gif Expert APP</h1>
            
            <AddCategory
                // sC={setCategorias}
                onNewCategory={(e:string) => {setCategorias([e,...categoriasHook])}}
                cHK={categoriasHook}
            />

            <ol>{categoriasHook.map(valor => {return <GifGrid key={v4()} categoria={valor}/>})}</ol>

            {/*Lista de gifs*/}
                {/*Gif cards*/}
        </>
    )
}