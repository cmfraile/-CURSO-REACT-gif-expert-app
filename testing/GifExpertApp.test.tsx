import { render , screen , fireEvent , waitFor, renderHook } from "@testing-library/react";
import { main as GifExpertApp } from '../src/components/main';
import React from "react";
import { useFetchGifs } from "../src/hooks/useFetchGifs";


jest.mock('../src/components/main.sass',() => {})

const main = () => {

    const inputvalue:string = 'The bad guys'
    
    test('Screenshot',() => {
        expect(render(<GifExpertApp/>).container).toMatchSnapshot();
    });

    test('Debe de generar consulta si el input cumple los requisitos',() => {
        render(<GifExpertApp/>);
        const input:HTMLInputElement = screen.getByLabelText('inputcat');
        const form:HTMLFormElement = screen.getByLabelText('form');
        fireEvent.input(input,{target:{value:inputvalue}});
        fireEvent.submit(form);
        
        expect(screen.getByRole('heading',{name:'Cargando...'})).toBeTruthy();
        expect(input.value).toBe('');
    
    });

    test('No debe de generar 2 consultas del mismo tema',() => {
        render(<GifExpertApp/>)
        const input:HTMLInputElement = screen.getByLabelText('inputcat');
        const form:HTMLFormElement = screen.getByLabelText('form');
        fireEvent.input(input,{target:{value:inputvalue}});
        fireEvent.submit(form);
        fireEvent.input(input,{target:{value:inputvalue}});
        fireEvent.submit(form);

        expect(screen.getAllByRole('heading',{name:'Cargando...'}).length).toBe(1);
    
    });

    test('No debe de generar consulta si el input no cumple los requisitos',() => {
        render(<GifExpertApp/>);
        const input:HTMLInputElement = screen.getByLabelText('inputcat');
        const form:HTMLFormElement = screen.getByLabelText('form');
        fireEvent.input(input,{target:{value:'a'}});
        expect(render(<GifExpertApp/>).container).toMatchSnapshot();
        fireEvent.submit(form);
        expect(render(<GifExpertApp/>).container).toMatchSnapshot();
    })

}

describe('Prueba de la ruta cr√≠tica de GifExpertApp',main)