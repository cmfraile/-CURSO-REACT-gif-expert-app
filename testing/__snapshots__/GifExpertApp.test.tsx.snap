// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de la ruta crítica de GifExpertApp No debe de generar consulta si el input no cumple los requisitos 1`] = `
<div>
  <h1>
    Gif Expert APP
  </h1>
  <form
    aria-label="form"
  >
    <input
      aria-label="inputcat"
      placeholder="buscar gifs"
      type="text"
      value=""
    />
  </form>
  <ol />
</div>
`;

exports[`Prueba de la ruta crítica de GifExpertApp No debe de generar consulta si el input no cumple los requisitos 2`] = `
<div>
  <h1>
    Gif Expert APP
  </h1>
  <form
    aria-label="form"
  >
    <input
      aria-label="inputcat"
      placeholder="buscar gifs"
      type="text"
      value=""
    />
  </form>
  <ol />
</div>
`;

exports[`Prueba de la ruta crítica de GifExpertApp Screenshot 1`] = `
<div>
  <h1>
    Gif Expert APP
  </h1>
  <form
    aria-label="form"
  >
    <input
      aria-label="inputcat"
      placeholder="buscar gifs"
      type="text"
      value=""
    />
  </form>
  <ol />
</div>
`;
